### PR TITLE
Validate array types when converting from variant contexts to internal rows

### PR DESCRIPTION
**Pull request by henrydavidge**
## What changes are proposed in this pull request?
Previously, when converting variant contexts to internal rows, we could throw confusing errors if the inner type of an array typed format field was not consistent between the expected schema and the variant context. Rather than receive a clear error that we expected type x for field y, the user would see a cryptic internal Spark error because of the type mismatch.

Now, we explicitly validate the elements in the variant context's collection type.

Note that compiler warnings could have tipped us off to the fact that we allowed bad values into the internal rows. As a future task, we should make compiler warnings (except probably deprecation warnings) fatal.

## How is this patch tested?
- [x] Unit tests
- [ ] Integration tests
- [x] Manual tests

(Details)

### CODE REVIEW

File: core/src/test/scala/io/projectglow/vcf/VariantContextToInternalRowConverterSuite.scala [hunk 1]
```diff
@@ -175,6 +175,33 @@ class VariantContextToInternalRowConverterSuite extends VCFConverterBaseTest {
...
+    def expectFailure(gp: Any): Unit = {
+      val vc = vcb.genotypes(new GenotypeBuilder().attribute("GP", gp).make()).make()
+      intercept[IllegalArgumentException] {
+        strictConverter.convertRow(vc, isSplit = false)
+      }
+    }
+
+    expectFailure(new JArrayList[Integer](Seq(1: Integer).asJavaCollection))
+    expectFailure(Array[Integer](1: Integer))
+    expectFailure(Array[Int](1))
```
Comment (karenfeng) [inline core/src/test/scala/io/projectglow/vcf/VariantContextToInternalRowConverterSuite.scala:193]: Let's also add a success case as a sanity check, and to clarify why these are failing.

File: core/src/test/scala/io/projectglow/vcf/VariantContextToInternalRowConverterSuite.scala [hunk 2]
```diff
@@ -175,6 +175,33 @@ class VariantContextToInternalRowConverterSuite extends VCFConverterBaseTest {
...
+  }
+
+  test("convert primitive array") {
+    val vcb = new VariantContextBuilder()
+    vcb.chr("").start(1).stop(1)
+    val refAllele = Allele.create("A", true)
+    vcb.alleles(Seq(refAllele).asJava)
+    val vc = vcb.genotypes(new GenotypeBuilder().attribute("GP", Array(1.0d)).make()).make()
+    strictConverter.convertRow(vc, isSplit = false)
```
Comment (karenfeng) [inline core/src/test/scala/io/projectglow/vcf/VariantContextToInternalRowConverterSuite.scala:202]: Should we check that this did what we expected?

File: core/src/main/scala/io/projectglow/vcf/VariantContextToInternalRowConverter.scala [hunk 3]
```diff
@@ -594,18 +595,32 @@ class VariantContextToInternalRowConverter(
...
+      primitiveConverter: Option[R => T] = None)(implicit ct: ClassTag[T]): Array[Any] =
     obj match {
       case null => null
       case VCFConstants.MISSING_VALUE_v4 => null
-      case arr: Array[T] => arr.asInstanceOf[Array[Any]]
-      case arr: Array[R] if primitiveConverter.isDefined => arr.map(primitiveConverter.get)
-      case l: JList[T] =>
+      case arr: Array[AnyRef] =>
+        var i = 0
+        while (i < arr.length) {
```
Comment (karenfeng) [inline core/src/main/scala/io/projectglow/vcf/VariantContextToInternalRowConverter.scala:606]: Do we have to check for every element, or is just the head enough?

File: core/src/main/scala/io/projectglow/vcf/VariantContextToInternalRowConverter.scala [hunk 4]
```diff
@@ -594,18 +595,32 @@ class VariantContextToInternalRowConverter(
...
+
+        arr.asInstanceOf[Array[Any]]
+      case arr: Array[_] if primitiveConverter.isDefined =>
+        arr.map(el => primitiveConverter.get(el.asInstanceOf[R]))
+      case l: JList[_] =>
         val arr = new Array[Any](l.size)
         var i = 0
         while (i < arr.length) {
           arr(i) = l.get(i)
+          require(
```
Comment (karenfeng) [inline core/src/main/scala/io/projectglow/vcf/VariantContextToInternalRowConverter.scala:621]: Same question here - do we have to check every element?

File: core/src/test/scala/io/projectglow/vcf/VariantContextToInternalRowConverterSuite.scala (hunk 1 above)
Comment (henrydavidge) [reply core/src/test/scala/io/projectglow/vcf/VariantContextToInternalRowConverterSuite.scala:191]: That's why I added the following test

File: core/src/main/scala/io/projectglow/vcf/VariantContextToInternalRowConverter.scala (hunk 3 above)
Comment (henrydavidge) [reply core/src/main/scala/io/projectglow/vcf/VariantContextToInternalRowConverter.scala:606]: It is possible for each element to have a different type if the original collection type was like `Array[Object]`. Knowing htsjdk, I wouldn't count against that.

Comment (karenfeng) [reply core/src/main/scala/io/projectglow/vcf/VariantContextToInternalRowConverter.scala:606]: OOC, does this add to the runtime significantly?

**Pull request merged**